### PR TITLE
🛠Consolidate JsonUtility and YamlUtility to share the same interface

### PR DESCRIPTION
**Pull request by yufeih**
There are some in-consistency between the public interface of `JsonUtility` and `YamlUtility`, this PR:

- Makes all `Parse` and `ToObject` method return an additional `error`, which means the input needs schema validation and any errors or warnings are reported through that return value.
- Makes all `DeserializeData` method turn the result directly, which means the input is a data file and we don't do validation for them, we'll throw if the data is corrupted.
- Removes `nullValidation` since they are always true (except for tests)

- Fix a TOC metadata validation bug caused by not calling the correct `ToObject` method
- Fix yaml header error does not have line info

@live1206 , do you know why this is changed? https://github.com/dotnet/docfx/pull/4194/files#diff-531ae800ae2cae4aa437350ad3816b00R102 It makes Yaml using `Deserialize` but leaves Json as `DeserializeWithSchemaValidation<JToken>`.

#4259

### CODE REVIEW

**Comment by live1206** [comment]
Both `docfx.yml` and `docfx.json` should use `Deserialize` instead of `DeserializeWithSchemaValidation`, I am changing it now while adding back line info

File: src/docfx/build/metadata/ContributionProvider.cs [hunk 1]
```diff
@@ -207,7 +207,7 @@ private async Task LoadCommitsTime(Docset docset)
             {
                 var (_, content, _) = await RestoreMap.GetRestoredFileContent(docset, docset.Config.Contribution.GitCommitsTime);
 
-                foreach (var commit in JsonUtility.Deserialize<GitCommitsTime>(content).Commits)
+                foreach (var commit in JsonUtility.DeserializeData<GitCommitsTime>(content).Commits)
```
Comment (OsmondJiang) [inline src/docfx/build/metadata/ContributionProvider.cs]: `DeserializeData` and `Deserialize` is not clear than previous `Deserialize` and `DeserializeWithSchemaValidation` from the method name.

Comment (yufeih) [reply src/docfx/build/metadata/ContributionProvider.cs]: Both are fine, as long as they are consistent, right now JToken parse with schema validation is also called `(Error, JToken) Deseriaize(string)`. I tend to want to make  `DeserializeWithSchemaValidation` the default behavior for docfx to detect more warnings, so called it `Deserialize`

Comment (live1206) [reply src/docfx/build/metadata/ContributionProvider.cs]: https://github.com/dotnet/docfx/blob/7e9e7db60909ad1ea80fe22d2a6b7a4e90029c0d/src/docfx/lib/YamlUtility.cs#L82
https://github.com/dotnet/docfx/blob/7e9e7db60909ad1ea80fe22d2a6b7a4e90029c0d/src/docfx/lib/YamlUtility.cs#L87
These two were missing during the last renaming

Comment (OsmondJiang) [reply src/docfx/build/metadata/ContributionProvider.cs]: From the usage of these two methods, seems `DeserializeData` is used more than `Deserialize`, apart from the tests.

File: test/docfx.Test/lib/JsonUtilityTest.cs [hunk 2]
```diff
@@ -47,7 +47,7 @@ public void TestObjectWithStringProperty(string input)
             var sw = new StringWriter();
             JsonUtility.Serialize(sw, new BasicClass { C = input });
             var json = sw.ToString();
-            var (errors, value) = JsonUtility.DeserializeWithSchemaValidation<BasicClass>(json);
+            var (errors, value) = Deserialize<BasicClass>(json);
```
Comment (OsmondJiang) [inline test/docfx.Test/lib/JsonUtilityTest.cs]: this Deserialize will cause confusion with JsonUtility/YamlUtility,Deserialize, rename it to DeserializeWithSchemaValidation?

**Pull request merged**